includes/data-port/models/class-sensei-data-port-model.php

```diff
@@ -0,0 +1,260 @@
+				$value = $config['default'];
+			}
+
+			if ( null !== $value ) {
+				switch ( $config['type'] ) {
+					case 'int':
+						$value = intval( $value );
+						break;
+					case 'float':
+						$value = floatval( $value );
+						break;
+					case 'bool':
+						$value = boolval( $value );
+						break;
+					case 'slug':
+						$value = sanitize_title( $value );
+						break;
+					case 'email':
+						$value = sanitize_email( $value );
+						break;
+					case 'url':
+						$value = esc_url_raw( $value );
+						break;
+					default:
+						if (
+							isset( $config['pattern'] )
+							&& 1 !== preg_match( $config['pattern'], $value )
+						) {
+							$value = null;
+						} elseif ( ! empty( $config['allow_html'] ) ) {
+							$value = wp_kses_post( $value );
+						} else {
+							$value = sanitize_text_field( $value );
+						}
+				}
+			}
+
+			$sanitized_data[ $key ] = $value;
+		}
+
+		$this->data = $sanitized_data;
+	}
+
+	/**
+	 * Get the post ID that this references.
+	 *
+	 * @return int
+	 */
+	public function get_post_id() {
+		return $this->post_id;
+	}
+
+	/**
+	 * Set the post ID that this references.
+	 *
+	 * @param int $id Post ID.
+	 */
+	protected function set_post_id( $id ) {
+		$this->post_id = $id;
+	}
+
+	/**
+	 * Get the data for the model.
+	 *
+	 * @return array
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Get the optional fields in the schema.
+	 *
+	 * @return array Field names.
+	 */
+	public static function get_optional_fields() {
+		$schema = static::get_schema();
+
+		return array_values(
+			array_filter(
+				array_map(
+					function( $field ) use ( $schema ) {
+						if ( empty( $schema[ $field ]['required'] ) ) {
+							return $field;
+						}
+
+						return false;
+					},
+					array_keys( $schema )
+				)
+			)
+		);
+	}
+
+	/**
+	 * Get the optional fields in the schema.
+	 *
+	 * @return array Field names.
+	 */
+	public static function get_required_fields() {
+		$schema = static::get_schema();
+
+		return array_values(
+			array_filter(
+				array_map(
+					function( $field ) use ( $schema ) {
+						if ( ! empty( $schema[ $field ]['required'] ) ) {
+							return $field;
+						}
+
+						return false;
+					},
+					array_keys( $schema )
+				)
+			)
+		);
+	}
+
+	/**
+	 * Get the schema for the data type.
+	 *
+	 * @return array {
+	 *     @type array $$field_name {
+	 *          @type string $type       Type of data. Options: string, int, float, bool, slug, ref, email, url.
+	 *          @type string $pattern    Regular expression that the value should match (Optional).
+	 *          @type mixed  $default    Default value if not set or invalid. Default is `null` (Optional).
+	 *          @type bool   $required   True if a non-empty value is required. Default is `false` (Optional).
+	 *          @type bool   $allow_html True if HTML should be allowed. Default is `false` (Optional).
+	 *     }
+	 * }
+	 */
+	public static function get_schema() {
```

Comment:
Why not define this as abstract?

Reply:
Static can't be abstract.

Reply:
Ah right! I now remember that I investigated this before (although I don't know why I was left with wrong impression). It seems that abstract static can be used in 7+ PHP only:
https://stackoverflow.com/a/31235907/12332297